### PR TITLE
Add host api validation and filtering

### LINKED ISSUE #1039: Host/ComputerSystem API support
**Is this a BUG REPORT or FEATURE REQUEST?**:

>
 /kind feature

**What happened**:

OpenSDS provides a platform for Data Management in on-premise as well as multi-cloud environment. Data Management includes identifying the resource. The node or Host is one of the key resources which manages as well as can facilitate the data in the data path.
OpenSDS should be aware of the Hosts it is managing. That means OpenSDS should be able to discover the Host and its associated resources. These information can be used for Host based replication, Host resource provisioning etc.

In addition, Issue #1023 also tell us it is essential to add host resource management in hotpot.

**What you expected to happen**:

1. Add host management feature.
2. Update attachment API based on new Host API.
3. Host management in Dashboard: https://github.com/opensds/opensds-dashboard/issues/205

**Anything else we need to know?**:

Tentative design at: [Click here](https://docs.google.com/document/d/1VoAHlZwQIubNCXCGeFdZjnXLSq7iqChHF_Ir6Y07o1E/edit)  

**Environment**:
- Hotpot(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by wisererik**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1039 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -217,6 +218,14 @@ func IsEqual(key string, value interface{}, reqValue interface{}) (bool, error)
 	}
 }
 
+// ToUpperFirst will change the first letter of string to uppercase
```
Comment (ThisIsClark) [inline pkg/utils/utils.go]: Could strings.Title meet your requirements？

Comment (wisererik) [reply pkg/utils/utils.go]: thanks, I change it to strings.Title, done

File: pkg/api/controllers/host.go [hunk 2]
```diff
@@ -76,7 +83,18 @@ func (p *HostPortal) CreateHost() {
 		return
 	}
 
-	// TODO: Add prameter validation
+	// HostName should be unique in the system
```
[the remark this answers is not in the document]
Comment (wisererik) [reply pkg/api/controllers/host.go]: thanks, if many places use it, we can create a function

[the remark this answers is not in the document]
Comment (ThisIsClark) [reply pkg/api/controllers/host.go]: ok

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -217,6 +221,79 @@ func IsEqual(key string, value interface{}, reqValue interface{}) (bool, error)
...
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			return x.Int() > y.Int()
+		}
+		return fmt.Sprintf("%v", x) > fmt.Sprintf("%v", y)
+	})
+	return arr
+}
+
+// Slicing implements pagination
+func Slice(arr interface{}, offset, limit int) interface{} {
```
Comment (ThisIsClark) [inline pkg/utils/utils.go]: The type of returned variables is `[]interface{}`, but the return type of this funcion is `interface{}`.
It looks weired, but I'm not sure whether it is what you want.

Comment (wisererik) [reply pkg/utils/utils.go]: interface{} includes []interface{}. it support developer to do like this: Slice(Sort(arr, sortKey, sortDir), offset, limit). 
if the actual kind of `data` is []interface{}.  we can convert it by `data.([]interface{})`

File: pkg/api/controllers/host.go [hunk 4]
```diff
@@ -41,8 +41,15 @@ func (p *HostPortal) ListHosts() {
 	if !policy.Authorize(p.Ctx, "host:list") {
 		return
 	}
-	// TODO: handle query parameters
-	hosts, err := db.C.ListHosts(c.GetContext(p.Ctx))
+
+	m, err := p.GetParameters()
+	if err != nil {
+		errMsg := fmt.Sprintf("list hosts failed: %s", err.Error())
```
Comment (kumarashit) [inline pkg/api/controllers/host.go]: This message should be changed. We are getting the parameters

File: pkg/api/controllers/host.go [hunk 5]
```diff
@@ -76,7 +83,18 @@ func (p *HostPortal) CreateHost() {
 		return
 	}
 
-	// TODO: Add prameter validation
+	// HostName should be unique in the system
+	hostArr, err := db.C.ListHostsByName(c.GetContext(p.Ctx), host.HostName)
+	if err != nil {
+		errMsg := fmt.Sprintf("list hosts failed in CreateHost method: %v", err)
```
Comment (kumarashit) [inline pkg/api/controllers/host.go]: You can print the hostName

File: pkg/api/controllers/host.go [hunk 6]
```diff
@@ -164,10 +182,27 @@ func (p *HostPortal) DeleteHost() {
...
-	// TODO: Add deletion constraition check
+	// Check relationship with volume
+	attachments, err := db.C.ListVolumeAttachmentsWithFilter(c.GetContext(p.Ctx), map[string][]string{"hostId": []string{id}})
+	if err != nil {
+		errMsg := fmt.Sprintf("list attachments failed in DeleteHost method: %v", err)
+		p.ErrorHandle(model.ErrorBadRequest, errMsg)
+		return
+	}
+	if len(attachments) > 0 {
+		errMsg := fmt.Sprintf("some volumes are attached to host: %s, please remove them first", host.HostName)
```
Comment (kumarashit) [inline pkg/api/controllers/host.go]:
```suggestion
		errMsg := fmt.Sprintf("some volumes are attached to host: %s, please detach them first", host.HostName)
```

File: pkg/utils/utils_test.go [hunk 7]
```diff
@@ -224,3 +225,278 @@ func TestIsEqual(t *testing.T) {
 		t.Errorf("Expected %v, get %v", expectedErr, err.Error())
 	}
 }
+
+type animal struct {
```
Comment (kumarashit) [inline pkg/utils/utils_test.go]: Animal ;)

Comment (wisererik) [reply pkg/utils/utils_test.go]: it's for unit testing only, so I use lowercase here

File: pkg/api/controllers/host.go (hunk 4 above)
Comment (wisererik) [reply pkg/api/controllers/host.go]: done

File: pkg/api/controllers/host.go (hunk 5 above)
Comment (wisererik) [reply pkg/api/controllers/host.go]: done

File: pkg/api/controllers/host.go (hunk 6 above)
Comment (wisererik) [reply pkg/api/controllers/host.go]: done

**Pull request merged**